scripts/merge-script/index.ts

```diff
@@ -0,0 +1,243 @@
+    const queue: PullRequest[] = [];
+    const allTargetBranches = new Set<string>();
+
+    for (const prNumber of pullRequests) {
+      const result = await loadAndValidatePullRequest(this, prNumber, force);
+
+      if (!isPullRequest(result)) {
+        failures.push(result);
+        continue;
+      }
+
+      // Push the determined target branches to the set of target branches. We keep
+      // track of all target branches so that we can fetch those before, and don't need
+      // to fetch the branches whenever an individual PR is getting merged.
+      result.targetBranches.forEach(branchName => allTargetBranches.add(branchName));
+      queue.push(result);
+    }
+
+    if (!queue.length) {
+      return {failures, status: MergeStatus.FAILED};
+    }
+
+    const fetchRefspecs: string[] = [];
+    const pushRefspecs: string[] = [];
+
+    for (const targetBranch of allTargetBranches) {
+      const localTargetBranch = this._getLocalTargetBranchName(targetBranch);
+      fetchRefspecs.push(`refs/heads/${targetBranch}:${localTargetBranch}`);
+      pushRefspecs.push(`${localTargetBranch}:refs/heads/${targetBranch}`);
+    }
+
+    // Fetch all target branches initially. We don't want to fetch them for each
+    // individual pull request as that causes an unnecessary slow-down.
+    if (this._gitGraceful(['fetch', '-f', this.repoGitUrl, ...fetchRefspecs]).status !== 0) {
+      return {status: MergeStatus.UNKNOWN_GIT_ERROR};
+    }
+
+    for (const pullRequest of queue) {
+      const failure = await this._mergePullRequest(pullRequest, dryRun);
+      // In case on pull request merge errors out critically, we abort the merge. We cannot
+      // proceed with the merge of other pull requests since we want to merge all specified
+      // pull requests together (respecting the order).
+      if (failure !== null) {
+        return {status: MergeStatus.FAILED, failures: [...failures, failure]};
+      }
+    }
+
+    // Delete all temporary branches. We don't want to keep the local repository
+    // in a dirty state.
+    this._git(['branch', '-D', PR_HEAD_NAME]);
+    this._git(['branch', '-D', PR_BASE_NAME]);
+    allTargetBranches.forEach(b => this._git(['branch', '-D', this._getLocalTargetBranchName(b)]));
+
+    return {
+      merged: queue,
+      status: MergeStatus.SUCCESS,
+    };
+  }
+
+  /**
+   * Merges the specified pull request. Method requires the temporary target branch
+   * to be fetched already. We don't want to fetch the target branches per pull request
+   * merge as it causes unnecessary fetch requests.
+   * @returns A pull request failure or null in case of success.
+   */
+  private async _mergePullRequest(
+      {prNumber, targetBranches, requiredBaseSha, commitsAmount}: PullRequest,
+      dryRun: boolean): Promise<PullRequestFailure|null> {
+    let previousBranch: string|null = null;
+    try {
+      previousBranch = this._git(['rev-parse', '--abbrev-ref', 'HEAD']).stdout.trim();
```

Comment:
I'd add a comment like "Remember the branch we're on now so we can change back to it at the end"